src/library/scala/runtime/ClassValueCompat.scala

```diff
@@ -0,0 +1,49 @@
+/*
+ * Scala (https://www.scala-lang.org)
+ *
+ * Copyright EPFL and Lightbend, Inc.
+ *
+ * Licensed under Apache License 2.0
+ * (http://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package scala.runtime
+
+private[scala] abstract class ClassValueCompat[T] {
+
+  private def checkClassValueAvailability(): Boolean =
+    try {
+      Class.forName("java.lang.ClassValue", false, classOf[Object].getClassLoader)
+      true
+    } catch {
+      case _: ClassNotFoundException => false
+    }
+
+  private val instance = if (checkClassValueAvailability()) new JavaClassValue() else new FallbackClassValue()
+
+  private class JavaClassValue extends ClassValue[T] with ClassValueInterface {
+    override def computeValue(cls: Class[_]): T = ClassValueCompat.this.computeValue(cls)
+  }
+
+  private class FallbackClassValue extends ClassValueInterface {
+    override def get(cls: Class[_]): T = ClassValueCompat.this.computeValue(cls)
+
+    override def remove(cls: Class[_]): Unit = {}
+  }
+
+  private trait ClassValueInterface {
+    def get(cls: Class[_]): T
+
+    def remove(cls: Class[_]): Unit
+  }
+
+  def get(cls: Class[_]): T = instance.get(cls)
+
+  def remove(cls: Class[_]): Unit = instance.remove(cls)
+
+  protected def computeValue(cls: Class[_]): T
+
+}
```

Comment:
Do you think, it would be better to have `instance` in constructor?
```suggestion
private[scala] abstract class ClassValueCompat[T] private (instance: ClassValueInterface[T])
    extends ClassValueInterface[T] { self =>

  def this() = this {
    try { new JavaClassValue() }
    catch { case _: ClassNotFoundException => new FallbackClassValue() }
  }

  private class JavaClassValue extends ClassValue[T] with ClassValueInterface[T] {
    override def computeValue(cls: Class[_]): T = self.computeValue(cls)
  }

  private class FallbackClassValue extends ClassValueInterface[T] {
    override def get(cls: Class[_]): T = self.computeValue(cls)

    override def remove(cls: Class[_]): Unit = {}
  }

  def get(cls: Class[_]): T = instance.get(cls)

  def remove(cls: Class[_]): Unit = instance.remove(cls)

  protected def computeValue(cls: Class[_]): T

}

private[scala] object ClassValueCompat {

  trait ClassValueInterface[T] {
    def get(cls: Class[_]): T
    def remove(cls: Class[_]): Unit
  }

}
```

Reply:
I am not exactly sure about the need of constructor, in which cases do we benefit from having it?

Reply:
* Having `instance` in constructor seemed conceptually simpler to me
 * `catch { case _: ClassNotFoundException => new FallbackClassValue() }` seemed like an elegant way to construct the object, while handling the possibility of `ClassValue` missing
 * I like when all `ClassValueCompat`, `JavaClassValue` and `FallbackClassValue` implement the same interface `ClassValueInterface`